Index: include/rcmail.php
===================================================================
diff --git a/trunk/program/include/rcmail.php b/trunk/program/include/rcmail.php
--- a/trunk/program/include/rcmail.php	(revision 49)
+++ b/trunk/program/include/rcmail.php	(working copy)
@@ -94,7 +94,11 @@
         }
 
         // start session
-        $this->session_init();
+        if (empty($GLOBALS['NOSESSION'])) {
+            if (!isset($_GET['_nosession'])) {
+                $this->session_init();
+            }
+        }
 
         // create user object
         $this->set_user(new rcube_user($_SESSION['user_id']));
@@ -724,7 +728,11 @@
     {
         $this->plugins->exec_hook('session_destroy');
 
-        $this->session->kill();
+        if ($this->session) {
+            $this->session->kill();
+            $_SESSION = array('language' => $this->user->language, 'temp' => true, 'skin' => $this->config->get('skin'));
+        }
+        
         $_SESSION = array('language' => $this->user->language, 'temp' => true, 'skin' => $this->config->get('skin'));
         $this->user->reset();
     }
Index: lib/Roundcube/rcube_db.php
===================================================================
diff --git a/trunk/program/lib/Roundcube/rcube_db.php b/trunk/program/lib/Roundcube/rcube_db.php
--- a/trunk/program/lib/Roundcube/rcube_db.php	(revision 49)
+++ b/trunk/program/lib/Roundcube/rcube_db.php	(working copy)
@@ -920,16 +920,22 @@
      *
      * @return string Translated table name
      */
-    public function table_name($table)
-    {
-        // add prefix to the table name if configured
-        if (($prefix = $this->options['table_prefix']) && strpos($table, $prefix) !== 0) {
-            return $prefix . $table;
-        }
+     public function table_name($table)
+     {
+         static $rcube;
 
-        return $table;
-    }
+         if (!$rcube) {
+             $rcube = rcube::get_instance();
+         }
 
+         // add prefix to the table name if configured
+         if (($prefix = $rcube->config->get('db_prefix')) && strpos($table, $prefix) !== 0) {
+             return $prefix . $table;
+         }
+
+         return $table;
+     }
+
     /**
      * Set class option value
      *
Index: lib/Roundcube/rcube_imap.php
===================================================================
diff --git a/trunk/program/lib/Roundcube/rcube_imap.php b/trunk/program/lib/Roundcube/rcube_imap.php
--- a/trunk/program/lib/Roundcube/rcube_imap.php	(revision 49)
+++ b/trunk/program/lib/Roundcube/rcube_imap.php	(working copy)
@@ -2327,6 +2327,8 @@
         }
 
         if ($saved) {
+            // let plugins know that a message has been appended
+            rcube::get_instance()->plugins->exec_hook("message_saved", array('saved' => $saved, 'folder' => $folder, 'message' => $message));
             // increase messagecount of the target folder
             $this->set_messagecount($folder, 'ALL', 1);
         }
Index: lib/Roundcube/rcube_imap_generic.php
===================================================================
diff --git a/trunk/program/lib/Roundcube/rcube_imap_generic.php b/trunk/program/lib/Roundcube/rcube_imap_generic.php
--- a/trunk/program/lib/Roundcube/rcube_imap_generic.php	(revision 49)
+++ b/trunk/program/lib/Roundcube/rcube_imap_generic.php	(working copy)
@@ -50,17 +50,17 @@
 
     public static $mupdate;
 
-    private $fp;
-    private $host;
-    private $logged = false;
-    private $capability = array();
-    private $capability_readed = false;
-    private $prefs;
-    private $cmd_tag;
-    private $cmd_num = 0;
-    private $resourceid;
-    private $_debug = false;
-    private $_debug_handler = false;
+    protected $fp;
+    protected $host;
+    protected $logged = false;
+    protected $capability = array();
+    protected $capability_readed = false;
+    protected $prefs;
+    protected $cmd_tag;
+    protected $cmd_num = 0;
+    protected $resourceid;
+    protected $_debug = false;
+    protected $_debug_handler = false;
 
     const ERROR_OK = 0;
     const ERROR_NO = -1;
@@ -352,7 +352,7 @@
      *
      * @return bool True if connection is closed
      */
-    private function eof()
+    protected function eof()
     {
         if (!is_resource($this->fp)) {
             return true;
@@ -375,7 +375,7 @@
     /**
      * Closes connection stream.
      */
-    private function closeSocket()
+    protected function closeSocket()
     {
         @fclose($this->fp);
         $this->fp = null;
@@ -421,7 +421,7 @@
         return false;
     }
 
-    private function hasCapability($name)
+    protected function hasCapability($name)
     {
         if (empty($this->capability) || $name == '') {
             return false;
@@ -1301,7 +1301,7 @@
      * @return array List of mailboxes or hash of options if $status_ops argument
      *               is non-empty.
      */
-    private function _listMailboxes($ref, $mailbox, $subscribed=false,
+    protected function _listMailboxes($ref, $mailbox, $subscribed=false,
         $status_opts=array(), $select_opts=array())
     {
         if (!strlen($mailbox)) {
@@ -1962,7 +1962,7 @@
      *
      * @return bool True on success, False on failure
      */
-    private function modFlag($mailbox, $messages, $flag, $mod = '+')
+    protected function modFlag($mailbox, $messages, $flag, $mod = '+')
     {
         if ($mod != '+' && $mod != '-') {
             $mod = '+';
@@ -3656,7 +3656,7 @@
         return $result;
     }
 
-    private function _xor($string, $string2)
+    protected function _xor($string, $string2)
     {
         $result = '';
         $size   = strlen($string);
@@ -3675,7 +3675,7 @@
      *
      * @return string Space-separated list of flags
      */
-    private function flagsToStr($flags)
+    protected function flagsToStr($flags)
     {
         foreach ((array)$flags as $idx => $flag) {
             if ($flag = $this->flags[strtoupper($flag)]) {
@@ -3727,7 +3727,7 @@
     /**
      * CAPABILITY response parser
      */
-    private function parseCapability($str, $trusted=false)
+    protected function parseCapability($str, $trusted=false)
     {
         $str = preg_replace('/^\* CAPABILITY /i', '', $str);
 
@@ -3804,7 +3804,7 @@
      *
      * @since 0.5-stable
      */
-    private function debug($message)
+    protected function debug($message)
     {
         if (($len = strlen($message)) > self::DEBUG_LINE_LENGTH) {
             $diff    = $len - self::DEBUG_LINE_LENGTH;
Index: lib/Roundcube/rcube_vcard.php
===================================================================
diff --git a/trunk/program/lib/Roundcube/rcube_vcard.php b/trunk/program/lib/Roundcube/rcube_vcard.php
--- a/trunk/program/lib/Roundcube/rcube_vcard.php	(revision 49)
+++ b/trunk/program/lib/Roundcube/rcube_vcard.php	(working copy)
@@ -135,7 +135,10 @@
         $this->firstname    = $this->raw['N'][0][1];
         $this->middlename   = $this->raw['N'][0][2];
         $this->nickname     = $this->raw['NICKNAME'][0][0];
-        $this->organization = $this->raw['ORG'][0][0];
+        for ($i = 0; $i < 3; $i++) {
+          $this->organization = $this->raw['ORG'][0][$i];
+          if ($this->organization) break;
+        }
         $this->business     = ($this->raw['X-ABSHOWAS'][0][0] == 'COMPANY') || (join('', (array)$this->raw['N'][0]) == '' && !empty($this->organization));
 
         foreach ((array)$this->raw['EMAIL'] as $i => $raw_email) {
@@ -525,7 +528,7 @@
     {
         // convert Apple X-ABRELATEDNAMES into X-* fields for better compatibility
         $vcard = preg_replace_callback(
-            '/item(\d+)\.(X-ABRELATEDNAMES)([^:]*?):(.*?)item\1.X-ABLabel:(?:_\$!<)?([\w-() ]*)(?:>!\$_)?./s',
+            '/item(\d+)\.(X-ABRELATEDNAMES)([^:]*?):(.*?)item\1.X-ABLabel:(?:_\$!<)?([\w-() ]*)(?:>!\$_)?./si',
             array('self', 'x_abrelatednames_callback'),
             $vcard);
 
@@ -532,9 +535,9 @@
         // Cleanup
         $vcard = preg_replace(array(
                 // convert special types (like Skype) to normal type='skype' classes with this simple regex ;)
-                '/item(\d+)\.(TEL|EMAIL|URL)([^:]*?):(.*?)item\1.X-ABLabel:(?:_\$!<)?([\w-() ]*)(?:>!\$_)?./s',
-                '/^item\d*\.X-AB.*$/m',  // remove cruft like item1.X-AB*
-                '/^item\d*\./m',         // remove item1.ADR instead of ADR
+                '/item(\d+)\.(TEL|EMAIL|URL)([^:]*?):(.*?)item\1.X-ABLabel:(?:_\$!<)?([\w-() ]*)(?:>!\$_)?./si',
+                '/^item\d*\.X-AB.*$/mi',  // remove cruft like item1.X-AB*
+                '/^item\d*\./mi',         // remove item1.ADR instead of ADR
                 '/\n+/',                 // remove empty lines
                 '/^(N:[^;\R]*)$/m',      // if N doesn't have any semicolons, add some
             ),
Index: localization/de_DE/messages.inc
===================================================================
diff --git a/trunk/program/localization/de_DE/messages.inc b/trunk/program/localization/de_DE/messages.inc
--- a/trunk/program/localization/de_DE/messages.inc	(revision 49)
+++ b/trunk/program/localization/de_DE/messages.inc	(working copy)
@@ -103,8 +103,8 @@
 $messages['messageopenerror'] = 'Die Nachricht konnte nicht vom Server geladen werden.';
 $messages['fileuploaderror'] = 'Hochladen der Datei fehlgeschlagen.';
 $messages['filesizeerror'] = 'Die hochzuladende Datei überschreitet die Maximalgröße von $size.';
-$messages['copysuccess'] = 'Erfolgreich kopiert $ nr Kontakte.';
-$messages['movesuccess'] = 'Erfolgreich verschoben $ nr Kontakte.';
+$messages['copysuccess'] = 'Erfolgreich kopiert $nr Kontakte.';
+$messages['movesuccess'] = 'Erfolgreich verschoben $nr Kontakte.';
 $messages['copyerror'] = 'Die Kontakte konnten nicht kopiert werden.';
 $messages['moveerror'] = 'Die Kontakte konnten nicht verschoben werden.';
 $messages['sourceisreadonly'] = 'Das Adressverzeichnis kann nicht verändert werden.';
Index: steps/addressbook/func.inc
===================================================================
diff --git a/trunk/program/steps/addressbook/func.inc b/trunk/program/steps/addressbook/func.inc
--- a/trunk/program/steps/addressbook/func.inc	(revision 49)
+++ b/trunk/program/steps/addressbook/func.inc	(working copy)
@@ -804,6 +804,9 @@
         $ff_value = '-del-'; // will disable delete-photo action
 
     $img = html::img(array('src' => $photo_img, 'border' => 1, 'alt' => ''));
+    
+    $RCMAIL->output->add_script('$("<img />").error(function() { $("#contactpic").children().attr("src", "skins/' . $RCMAIL->config->get('skin', 'larry') . '/images/contactpic.png") }).attr("src", "' . $photo_img . '");', 'foot');
+    
     $content = html::div($attrib, $img);
 
     if ($CONTACT_COLTYPES['photo'] && ($RCMAIL->action == 'edit' || $RCMAIL->action == 'add')) {
Index: steps/addressbook/photo.inc
===================================================================
diff --git a/trunk/program/steps/addressbook/photo.inc b/trunk/program/steps/addressbook/photo.inc
--- a/trunk/program/steps/addressbook/photo.inc	(revision 49)
+++ b/trunk/program/steps/addressbook/photo.inc	(working copy)
@@ -67,6 +67,11 @@
 // let plugins do fancy things with contact photos
 $plugin = $RCMAIL->plugins->exec_hook('contact_photo',
     array('record' => $record, 'email' => $email, 'data' => $data));
+    
+// check if data is an URL
+if (filter_var($data, FILTER_VALIDATE_URL)) {
+    $plugin['url'] = $data;
+}
 
 // redirect to url provided by a plugin
 if ($plugin['url']) {
Index: steps/mail/show.inc
===================================================================
diff --git a/trunk/program/steps/mail/show.inc b/trunk/program/steps/mail/show.inc
--- a/trunk/program/steps/mail/show.inc	(revision 49)
+++ b/trunk/program/steps/mail/show.inc	(working copy)
@@ -350,6 +350,8 @@
     else {
         $photo_img = $placeholder ? $placeholder : 'program/resources/blank.gif';
     }
-
+    
+    $RCMAIL->output->add_script('$("<img />").error(function() { $("#contactphoto").children().remove() }).attr("src", "' . $photo_img . '");', 'foot');
+    
     return html::img(array('src' => $photo_img) + $attrib);
 }
Index: steps/settings/about.inc
===================================================================
diff --git a/trunk/program/steps/settings/about.inc b/trunk/program/steps/settings/about.inc
--- a/trunk/program/steps/settings/about.inc	(revision 49)
+++ b/trunk/program/steps/settings/about.inc	(working copy)
@@ -50,7 +50,8 @@
         $attrib['id'] = 'rcmpluginlist';
     }
 
-    $plugins     = array_filter((array) $RCMAIL->config->get('plugins'));
+    $plugins = (array) $RCMAIL->config->get('plugins');
+    $plugins = array_filter($RCMAIL->plugins->exec_hook('plugins_installed', $plugins));
     $plugin_info = array();
 
     foreach ($plugins as $name) {
